uaclient/api/u/pro/attach/auto/full_auto_attach/v1.py

```diff
@@ -0,0 +1,168 @@
+    contract_client = contract.UAContractClient(cfg)
+    try:
+        tokenResponse = contract_client.request_auto_attach_contract_token(
+            instance=cloud
+        )
+    except exceptions.ContractAPIError as e:
+        if e.code and 400 <= e.code < 500:
+            raise exceptions.NonAutoAttachImageError(
+                messages.UNSUPPORTED_AUTO_ATTACH
+            )
+        raise e
+
+    token = tokenResponse["contractToken"]
+    return token
+
+
+def _is_any_beta(cfg: UAConfig, ents: List[str]) -> Tuple[bool, str]:
+    for name in ents:
+        try:
+            ent_cls = entitlements.entitlement_factory(cfg, name)
+            if ent_cls.is_beta:
+                return True, name
+        except exceptions.EntitlementNotFoundError:
+            continue
+    return False, ""
+
+
+def _is_incompatible_services_present(
+    cfg, ent_list: List[str]
+) -> Tuple[bool, str, List[str]]:
+    ent_cls = list()  # type: List[Type[UAEntitlement]]
+    for e in ent_list:
+        ent_cls.append(entitlements.entitlement_factory(cfg, e))
+
+    for ent in ent_cls:
+        ent_inst = ent(cfg)
+        if ent_inst.incompatible_services:
+            incompat = [
+                service.entitlement
+                for service in ent_inst.incompatible_services
+            ]
+            if any(e in incompat for e in ent_cls):
+                return True, ent_inst.name, [e(cfg).name for e in incompat]
+    return False, "", []
+
+
+def full_auto_attach(options: FullAutoAttachOptions) -> FullAutoAttachResult:
+    return _full_auto_attach(options, UAConfig(root_mode=True))
+
+
+def _full_auto_attach(options: FullAutoAttachOptions, cfg: UAConfig):
+    event.set_event_mode(event_logger.EventLoggerMode.JSON)
+
+    services = set()
+    if options.enable:
+        is_beta_found, service = _is_any_beta(cfg, options.enable)
+        if is_beta_found:
+            raise exceptions.BetaServiceError(
+                msg="beta service found in the enable list",
+                msg_code="beta-service-found",
+                additional_info={"beta_service": service},
+            )
+        services.update(options.enable)
+    if options.enable_beta:
+        services.update(options.enable_beta)
+
+    service_list = list(services)
+
+    found, not_found = entitlements.get_valid_entitlement_names(
+        service_list, cfg
+    )
+    if not_found:
+        msg = entitlements.create_enable_entitlements_not_found_message(
+            not_found, cfg=cfg, allow_beta=True
+        )
+        raise exceptions.EntitlementNotFoundError(msg.msg, not_found)
+
+    incompat_detected, ent, incompat_ents = _is_incompatible_services_present(
+        cfg, sorted(found)
+    )  # sort for easy testing exceptions raised
+    if incompat_detected:
+        msg_ = "{ent} is incompatible with any of these services {incompat}"
+        raise exceptions.IncompatibleEntitlementsDetected(
+            msg=msg_.format(ent=ent, incompat=incompat_ents),
```

Comment:
why is there an extra underscore in this variable name?
```suggestion
        msg = "{ent} is incompatible with any of these services {incompat}"
        raise exceptions.IncompatibleEntitlementsDetected(
            msg=msg.format(ent=ent, incompat=incompat_ents),
```

Reply:
mypy has errors. It thinks `msg` is of type `NamedMessage` because it is first created 
```
entitlements.create_enable_entitlements_not_found_message(
            not_found, cfg=cfg, allow_beta=True
        )
```